tests/chainer_tests/functions_tests/math_tests/test_polygamma.py

```diff
@@ -1,81 +1,60 @@
-from chainer import gradient_check
 from chainer import testing
-from chainer.testing import attr
 
 
 @testing.parameterize(*testing.product({
     'shape': [(), (3, 2)],
     'dtype': [numpy.float16, numpy.float32, numpy.float64]
 }))
 @testing.with_requires('scipy')
-class TestPolyGamma(unittest.TestCase):
+@testing.inject_backend_tests(
+    None,
+    # CPU tests
+    [
+        {},
+    ]
+    # GPU tests
+    + testing.product({
+        'use_cuda': [True],
+        'cuda_device': [0, 1],
+    })
+    # ChainerX tests
+    + testing.product({
+        'use_chainerx': [True],
+        'chainerx_device': ['native:0', 'cuda:0', 'cuda:1']
+    })
+)
+class TestPolyGamma(testing.FunctionTestCase):
 
     def setUp(self):
-        self.x = \
-            numpy.random.uniform(1., 10., self.shape).astype(self.dtype)
-        self.gy = numpy.random.uniform(-1, 1, self.shape).astype(self.dtype)
-        self.ggx = \
-            numpy.random.uniform(-1, 1, self.shape).astype(self.dtype)
-        self.n = numpy.random.randint(3, size=self.shape).astype(numpy.int32)
         if self.dtype == numpy.float16:
             self.check_forward_options = {'atol': 1e-2, 'rtol': 1e-2}
         else:
             self.check_forward_options = {'atol': 1e-4, 'rtol': 1e-4}
-        self.check_backward_options = {'eps': 1e-3, 'atol': 5e-2, 'rtol': 1e-3,
-                                       'dtype': numpy.float64}
+        self.check_backward_options = {'eps': 1e-3, 'atol': 5e-2, 'rtol': 1e-3}
         self.check_double_backward_options = {'eps': 1e-3, 'atol': 5e-2,
-                                              'rtol': 1e-3,
-                                              'dtype': numpy.float64}
-
-    def check_forward(self, n_data, x_data):
-        from scipy import special
-        x = chainer.Variable(x_data)
-        n = chainer.Variable(n_data)
-        y = F.polygamma(n, x)
-        testing.assert_allclose(
-            special.polygamma(self.n, self.x), y.data,
-            **self.check_forward_options)
+                                              'rtol': 1e-3}
 
-    def test_polygamma_forward_cpu(self):
-        self.check_forward(self.n, self.x)
-
-    @attr.gpu
-    def test_polygamma_forward_gpu(self):
-        self.check_forward(cuda.to_gpu(self.n), cuda.to_gpu(self.x))
-
-    def check_backward(self, n_data, x_data, y_grad):
-        gradient_check.check_backward(
-            lambda x: F.polygamma(chainer.Variable(n_data), x), x_data, y_grad,
-            **self.check_backward_options)
+    def generate_inputs(self):
+        n = numpy.random.randint(3, size=self.shape).astype(numpy.int32)
+        x = numpy.random.uniform(1., 10., self.shape).astype(self.dtype)
+        return n, x
 
-    def test_polygamma_backward_cpu(self):
-        self.check_backward(self.n, self.x, self.gy)
+    def forward_expected(self, inputs):
+        n, x = inputs
+        y_expect = scipy.special.polygamma(n, x)
+        return y_expect.astype(self.dtype),
```

Comment:
Do we get test failures without this cast?

Reply:
yes. `dtypes do not match` error

Reply:
It's not obvious in what cases so I think it's better to leave a comment in the code when/why it happens.